Fixes #38391 - Allow displaying all host interfaces through API

adamlazik1 opened the pull request:
Currently, when you add an interfce to an existing host, for example
with
```
hammer host interface create --host <hostname> --identifier test --managed false
```
and the try to GET request to display the interfaces with
```
curl --request GET --insecure --user admin:changeme https://`hostname -f`/api/v2/hosts/10/enc
```
you get an error about undefined method `to_export`. For details see the
linked issue.

The default class of the interface without provided type seems to be
`Nic::Base`. We need to provide a valid interface type as a default to
ensure that every interface has a valid type and can be exported.

adamlazik1 committed:
Fixes #38391 - Allow displaying all host interfaces through API

Currently, when you add an interfce to an existing host, for example
with
```
hammer host interface create --host <hostname> --identifier test --managed false
```
and the try to GET request to display the interfaces with
```
curl --request GET --insecure --user admin:changeme https://`hostname -f`/api/v2/hosts/10/enc
```
you get an error about undefined method `to_export`. For details see the
linked issue.

The default class of the interface without provided type seems to be
`Nic::Base`. We need to provide a valid interface type as a default to
ensure that every interface has a valid type and can be exported.

ofedoren reviewed (commented):
I think the underlying issue is more complex than just allowing things. It seems `Nic::Base` should not be created at all. The least usable class should be `Nic::Interface`. There are some points suggesting/confirming that:
 - There was an idea of making `Nic::Base` an abstract class: https://github.com/theforeman/foreman/pull/9547, you can find more info in the attached RM/BZ issue.
 - API description says that the default type (if none provided I assume) should be `Nic::Managed`: https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/interfaces_controller.rb#L35 / https://github.com/theforeman/foreman/blob/develop/app/services/interface_type_mapper.rb#L4
  - in UI there is no option such as `Base`, but `Interface`, `BMC`, etc
  - in API controller we allow nil type (https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hosts_controller.rb#L365) for new hosts, but it still gets converted into `Nic::Managed`: https://github.com/theforeman/foreman/blob/develop/app/services/interface_type_mapper.rb#L9
   - `Nic::Base.allowed_types` returns `Nic::Managed`, `Nic::BMC`, `Nic::Bond` and `Nic::Bridge` suggesting that even `Nic::Interface` should not be created.

adamlazik1 commented:
@ofedoren thank you for the feedback. I found out that methods in hosts controller allow nil interface type, which seems strange to me. The issue is fixed when I hardcode in the default type. I wonder if we should go with this approach or instead make `type` a required parameter. I suspect that the latter could possibly break workflows of some users. There is also question of what we should do with existing nil type interfaces. Would it be safe to change the type with a migration?

ofedoren commented on db/migrate/20250527151609_change_base_interface_type.rb:
```diff
@@ -0,0 +1,12 @@
+      UPDATE nics
+      SET type = 'Nic::Managed'
+      WHERE type IS NULL;
+    SQL
```
What's the benefit of using raw SQL here instead of Rails' ORM?

adamlazik1 replied:
Well, it was throwing a bunch of errors like this at me:
```
`method_missing': undefined method `pulp_primary' for SmartProxy:Class (NoMethodError)
...
```
But when I ran the command in the console it worked. I am not sure if it's just an issue of my dev setup. Anyway I updated the migration to ORM, please double check that it is correct.

ofedoren commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -370,7 +370,7 @@ def host_attributes(params, host = nil)
 
       def interface_attributes(params, allow_nil_type: false)
-        params[:type] = InterfaceTypeMapper.map(params[:type]) if params.has_key?(:type) || allow_nil_type
+        params[:type] = InterfaceTypeMapper.map(params[:type]) if params.has_key?(:type) || allow_nil_type || (params[:type].nil? && params[:id].nil?)
```
Does the new condition make sense?

`params[:type].nil?` will only be checked if `params.has_key?(:type)` is `false`, which implicitly means that the type is `nil`, since it's not even in the hash.

`params[:id].nil`? will only be checked if `allow_nil_type` is `false`, which means the action is `update`, thus the host id must be present making `params[:id].nil?` always return `false`.

adamlazik1 replied:
> params[:type].nil? will only be checked if params.has_key?(:type) is false, which implicitly means that the type is nil, since it's not even in the hash.

True,  I didn't think of that. I will fix it.

> params[:id].nil? will only be checked if allow_nil_type is false, which means the action is update, thus the host id must be present making params[:id].nil? always return false.

In this method params[:id] refers to the interface id instead of the host id. Such request is created for example with `hammer host interface create`, which in reality sends a PUT request to the host controller.

adamlazik1 replied:
Anyway my aim is that it behaves exactly the same as the previous version of this PR:
```ruby
params[:type] = InterfaceTypeMapper.map(params[:type]) if params.has_key?(:type) || allow_nil_type
params[:type] ||= 'Nic::Managed' if params[:id].nil?
```

Except the fact that I let the InterfaceTypeMapper decide the default interface, same way it is done in interfaces controller.

ofedoren commented on db/migrate/20250527151609_change_base_interface_type.rb:
```diff
@@ -0,0 +1,8 @@
+class ChangeBaseInterfaceType < ActiveRecord::Migration[7.0]
+  def up
+    Nic::Base.where(type: nil).update_all(type: 'Nic::Managed')
```
I guess we need to convert `Nic::Base` or other possible types which are not in `allowed` ones?

adamlazik1 replied:
Nic::Base in this case is just a representation for the nics table in the database, I believe it effectively encapsulates all of the interfaces.

ofedoren replied:
What I meant is to check if `type` is not just `nil`, but any of the types which are not considered in `Nic::Base.allowed_types`, e.g. `Nic::Interface`.

Manually I can create an interface with `Nic::Interface` type and attach it to a host. If we want to add a clean up migration, we'd probably also want to clean any value we don't expect/accept. Although, we still don't have a strong check, so we'll hope that wrong values won't get saved by an accident or manual intruding.

@adamruzicka, any opinion on this?

adamlazik1 replied:
Ah, I misunderstood. Updated.

It is my hope that we don't have to account for manual meddling of the database, consequences of that should become the user's responsibility. And I believe we have a check for unknown interface types other than nil, so we should be covered from that point.
```
[vagrant@fdev hammer-cli]$ hm host interface create --host-id 3 --type unknown  --identifier interface
Could not create the interface:
  Error: Option '--type': Value must be one of 'interface', 'bmc', 'bond', 'bridge'..
  
  See: 'hammer host interface create --help'.
[vagrant@fdev hammer-cli]$ 
```

ofedoren reviewed (commented):
Thanks, @adamlazik1, but I still think there is something off:

https://github.com/theforeman/foreman/pull/4833 was meant to fix an issue where we allowed to update the type of already created interface. This should remain. But at the same time it introduced `allow_nil_type`, which we are now trying to forbid 'cause it leads to either `nil` or `Nic::Base` type. With this PR we are going to make `Nic::Managed` to be the actual default type. I'd say we need to drop `allow_nil_type`.
 
 `allow_nil_type` is essentially `host.nil?`, which seems to mean that we allow nil type in case the host is just being created. But even in the API docs we say it will be the default type if not provided :shrug:.
 
 Since we won't allow neithier nil nor base, we also need to update this line: https://github.com/theforeman/foreman/blob/develop/app/models/nic/base.rb#L349
 
Could we also add a test to simulate the issue is or at least is described in the RM ticket?

adamlazik1 commented:
> Thanks, @adamlazik1, but I still think there is something off:
> 
> #4833 was meant to fix an issue where we allowed to update the type of already created interface. This should remain. But at the same time it introduced `allow_nil_type`, which we are now trying to forbid 'cause it leads to either `nil` or `Nic::Base` type. With this PR we are going to make `Nic::Managed` to be the actual default type. I'd say we need to drop `allow_nil_type`.
> 
> `allow_nil_type` is essentially `host.nil?`, which seems to mean that we allow nil type in case the host is just being created. But even in the API docs we say it will be the default type if not provided 🤷.
> 
> Since we won't allow neithier nil nor base, we also need to update this line: https://github.com/theforeman/foreman/blob/develop/app/models/nic/base.rb#L349
> 
> Could we also add a test to simulate the issue is or at least is described in the RM ticket?

Sure, I removed allow_nil_type and and added a test to ensure that a default interface type is being assigned.

ofedoren commented on app/models/nic/base.rb:
```diff
@@ -346,7 +346,7 @@ def validate_mac_is_unicast
     def validate_updating_types
-      sti_type = type || 'Nic::Base'
+      sti_type = type
       errors.add(:type, _("can't be changed once the interface is saved")) if persisted? && (self.class.name != sti_type)
```
Isn't this the same as:

```suggestion
      errors.add(:type, _("can't be changed once the interface is saved")) if persisted? && (self.class.name != type)
```

adamlazik1 replied:
It is, truth be told I don't know why I left it like this. Applied.

ofedoren commented on test/controllers/api/v2/hosts_controller_test.rb:
```diff
@@ -317,6 +317,16 @@ def last_record
     assert_equal Nic::BMC,     last_record.interfaces.find_by_mac('00:11:22:33:44:01').class
   end
 
+  test "should provide a default type for interfaces" do
```
This is nice. I think there already should be a test to ensure we forbid type changing (through API) for already saved interfaces, could you please re-check?.. If not then we'd need one.

adamlazik1 replied:
Yes, we do have this test on line 413:
```ruby
  test "updating interface type isn't allowed" do
    @host = FactoryBot.create(:host, :interfaces => [FactoryBot.build(:nic_bond, :primary => true)])
    nic_id = @host.interfaces.first.id
    put :update, params: { :id => @host.to_param, :host => { :interfaces_attributes => [{ :id => nic_id, :name => 'newname', :type => 'bmc'}] } }

    assert_response :unprocessable_entity
    body = ActiveSupport::JSON.decode(response.body)
    assert_includes body['error']['errors'].keys, 'interfaces.type'
  end
```

ofedoren reviewed (commented):
Thanks, @adamlazik1, few nitpicks:

ofedoren reviewed (approved):
Thanks, @adamlazik1, AFAICT this looks nice and makes me happy, but I'm still not sure if I didn't miss anything, thus: @adamruzicka, any concerns?

adamruzicka reviewed (commented):
The actual changes seem to have drifted a bit from the commit message, apart from that this should be good to go

adamlazik1 commented:
Thank you for reminding me. Updated the commit message.

ofedoren reviewed (approved):
Thanks, @adamlazik1 and @adamruzicka !
